### ISSUE TITLE
Automatically build host-vm

### ISSUE THREAD

**Issue by MeFisto94** [ReadyTalk/hello-ios#7]
Hey Guys,
My Problem is this:
```
[exec] generating bootimage and codeimage binaries from build/ios-arm64-lzma-bootimage-openjdk-src/classpath using build/ios-arm64-lzma-bootimage-openjdk-src/bootimage-generator
     [exec] unable to open build/macosx-x86_64-interpret/libjvm.so
```
So this means if you cross-compile, the bootimage-generator simply expects it's host-machine to have a virtual machine previously compiled?

Actually the code already generated a `ios-arm64-interpret-lzma-openjdk-src` (Interpreted Version), but I have the impression that the make-file simply doesn't recognize it due to the many parameters and the way host-vm is set up. 
Also host-vm is defined but never really build by the Makefile. 

The next thing is that there is only a libjvm.dylib, since so is linux only. Adopting the Makefile leads to
```
[exec] unable to initialize host VM
```
Whereas the host VM is simply `JAVA_HOME=openjdkDir make process=interpret`

**Comment by dicej** [comment]
Hi,

Thanks for reporting this issue.  I just tried reproducing this problem, but couldn't.  However, I don't have a current valid signing identity, nor do I have a real iOS device to test with, so it might be that I'm not getting as far as you in the build process.  Here's what I did, in case it's useful:

    wget https://bitbucket.org/alexkasko/openjdk-unofficial-builds/downloads/openjdk-1.7.0-u80-unofficial-macosx-x86_64-image.zip
    hg clone http://hg.openjdk.java.net/jdk7u/jdk7u/
    cd jdk7u
    bash get_source.sh
    cd jdk
    hg checkout jdk7u80-b32
    cd ../../
    git clone git@github.com:readytalk/avian
    git clone git@github.com:readytalk/hello-ios
    curl -Of http://readytalk.github.io/avian-web/lzma920.tar.bz2
    (mkdir -p lzma-920 && cd lzma-920 && tar xjf ../lzma920.tar.bz2)
    curl -Of http://readytalk.github.io/avian-web/proguard4.11.tar.gz
    tar xzf proguard4.11.tar.gz
    cd hello-ios
    make openjdk=$(pwd)/../openjdk-1.7.0-u80-unofficial-macosx-x86_64-image/ openjdk-src=$(pwd)/../jdk7u/jdk/src/ lzma=$(pwd)/../lzma-920 run-proguard=true

At that point I ran into an issue with the avian build which I addressed here: https://github.com/ReadyTalk/avian/pull/494

Once that was fixed, the build continued successfully until the xcodebuild step, at which point it complained that I didn't have a valid signing identity, which is true (and I'm not inclined to pay the $99 required to register as an iOS dev right now, especially since I don't have an iOS device to test with anyway).

However, I was able to build and run the project for the simulator when I added the `sim=true` flag to the above `make` command.  I did have to fix a small issue in the ProGuard config, though: https://github.com/ReadyTalk/hello-ios/commit/d5e12c523a334a60ec1126c46badc801f8a31122

**Comment by joshuawarner32** [comment]
@MeFisto94: I think this PR might fix your issue: https://github.com/ReadyTalk/avian/pull/495

**Comment by MeFisto94** [comment]
My Apologies,
I just recognized that this Issue should rather belong to avian since it's completely unrelated to hello-ios actually (I am trying to generate an avian stub for the jMonkeyEngine so we can deploy iOS Apps again).

As for ReadyTalk/avian#494 I also ran into that issue and simply copied the files from /usr/include, but nice to see it working.

ReadyTalk/avian#495 didn't fix it for me though, fo rsome reason I only had the ios-arm-bootimage folder created, maybe it's some order differences?
My trouble is in `$(bootimage-object) $(codeimage-object): $(bootimage-generator) \` so it depends on bootimage-generator, though the other part states `$(bootimage-generator): $(bootimage-generator-objects) $(vm-objects)`, so the folders should've already been built?

**Comment by dicej** [comment]
Can you post a test case somewhere so we can reproduce it?  I'm guessing you're using Ant based on the `[exec]` lines in your original report, but it's hard to guess what's going wrong without seeing the build.xml, etc.

**Comment by MeFisto94** [comment]
I'll try to put up a test case, yes, though it's strange that you don't experience the same issues and that I get inconsistent results (the error where I'm stuck, I couldn't pass it before due to the ld: "cannot insert branch island", whereas arm64 yielded the error in bootimage-generation, now back to arm it does try to generate the bootimage)

First of all: Adding sim=true to the make command led to errors when trying to compile arm-asm.o on arch=asm. 
`JAVA_HOME=[...]/openjdk-1.7.0-u80-unofficial-macosx-x86_64-image/ make arch=arm platform=ios bootimage=true process=compile openjdk=[...]/openjdk-1.7.0-u80-unofficial-macosx-x86_64-image/ openjdk-src=[...]/jdk7u/jdk/src/ sim=true`

```
src/arm.S:37:3: error: unknown use of instruction mnemonic without a size suffix
  mov ip, sp
  ^
src/arm.S:38:11: error: unexpected token in argument list
  stmfd sp!, {r4-r8, lr}
          ^
src/arm.S:41:3: error: unknown use of instruction mnemonic without a size suffix
  mov r4, r0
  ^
src/arm.S:42:3: error: unknown use of instruction mnemonic without a size suffix
  mov r5, r1
  ^
src/arm.S:43:11: error: brackets expression not supported on this target
  ldr r6, [ip]
          ^
src/arm.S:44:11: error: brackets expression not supported on this target
  ldr r7, [ip, #4]
          ^
src/arm.S:45:11: error: brackets expression not supported on this target
  ldr r8, [ip, #8]
          ^
src/arm.S:48:3: error: unknown use of instruction mnemonic without a size suffix
  sub sp, sp, r5
  ^
src/arm.S:49:3: error: unknown use of instruction mnemonic without a size suffix
  mov ip, sp
  ^
src/arm.S:51:3: error: invalid instruction mnemonic 'tst'
  tst r3, r3
  ^~~
src/arm.S:52:13: error: brackets expression not supported on this target
  ldrne r0, [r2], #4
            ^
src/arm.S:53:13: error: brackets expression not supported on this target
  strne r0, [ip], #4
            ^
src/arm.S:54:17: error: unknown token in expression
  subne r3, r3, #4
                ^
src/arm.S:55:3: error: invalid instruction mnemonic 'bne'
  bne .Lloop
  ^~~
src/arm.S:58:3: error: invalid instruction mnemonic 'tst'
  tst r6, r6
  ^~~
src/arm.S:60:15: error: unknown token in expression
  ldmiane r6, {r0-r3}
              ^
src/arm.S:73:3: error: invalid instruction mnemonic 'blx'
  blx r4
  ^~~
src/arm.S:75:3: error: unknown use of instruction mnemonic without a size suffix
  add sp, sp, r5
  ^
src/arm.S:90:11: error: unexpected token in argument list
  ldmfd sp!, {r4-r8, pc}
          ^
src/arm.S:95:3: error: unknown use of instruction mnemonic without a size suffix
  mov lr, r0
  ^
src/arm.S:96:11: error: brackets expression not supported on this target
  ldr r0, [sp]
          ^
src/arm.S:97:11: error: brackets expression not supported on this target
  ldr r1, [sp, #4]
          ^
src/arm.S:98:3: error: unknown use of instruction mnemonic without a size suffix
  mov sp, r2
  ^
src/arm.S:99:3: error: unknown use of instruction mnemonic without a size suffix
  mov r8, r3
  ^
src/arm.S:100:3: error: invalid instruction mnemonic 'bx'
  bx lr
  ^~
src/arm.S:111:11: error: unexpected token in argument list
  stmfd sp!, {r4-r11, lr}
          ^
src/arm.S:113:15: error: unknown token in expression
  sub sp, sp, #12
              ^
src/arm.S:115:11: error: brackets expression not supported on this target
  str sp, [r2, #24]
          ^
src/arm.S:117:3: error: unknown use of instruction mnemonic without a size suffix
  mov r12, r0
  ^
src/arm.S:118:11: error: brackets expression not supported on this target
  ldr r0, [r2, #4]
          ^
src/arm.S:124:3: error: invalid instruction mnemonic 'blx'
  blx r12
  ^~~
src/arm.S:130:15: error: unknown token in expression
  add sp, sp, #12
              ^
src/arm.S:131:11: error: unexpected token in argument list
  ldmfd sp!, {r4-r11, lr}
          ^
src/arm.S:132:3: error: invalid instruction mnemonic 'bx'
  bx lr
  ^~
```
Stuff like the "mov without size" makes me think it might be a wrong-platform thing? Though I can't imagine where you need to specify a size because the assembler should auomatically determine that.

I'll try to setup a test case tomorrow, I don't know how to do it reproducable, though. Travis would be nice, or some docker container? But that's a bit of an overhead.

Essentially the build.xml simply calls make and packages a few output files to a zip for each platform (arm, arm64 and i386).

```ant
 <exec executable="make" dir="${avian.src.path}" failonerror="true">
    <env key="JAVA_HOME" value="${avian.jdk.path}"/>
    <arg line="arch=arm"/>
    <arg line="platform=ios"/>
    <arg line="process=compile"/>
    <arg line="openjdk='${avian.jdk.path}'"/>
    <arg line="openjdk-src='${avian.jdk.src.path}'"/>
    <arg line="lzma='${avian.lzma.sdk.path}'"/>
    <arg line="bootimage=true"/>
</exec>
<delete file="src/com/jme3/gde/ios/avian-openjdk-mac.zip" failonerror="false"/>
<zip destfile="src/com/jme3/gde/ios/avian-openjdk-mac.zip">
     <zipfileset file="${avian.src.path}/build/ios-arm-bootimage-openjdk-src/bootimage-generator" filemode="755" fullpath="avian-arm/bootimage-generator"/>
     <zipfileset file="${avian.src.path}/build/ios-arm-bootimage-openjdk-src/binaryToObject/binaryToObject" filemode="755" fullpath="avian-arm/binaryToObject"/>
     <zipfileset file="${avian.src.path}/build/ios-arm-bootimage-openjdk-src/libavian.a" fullpath="avian-arm/libavian.a"/>
</zip>
```

So it actually fails on the make part. I could work around many issues though (Missing GetClassLoader in jvm.h, so I took it from hotspot, missing netinet headers (btw, does ReadyTalk/avian#494 also take into account that there were some files from `netinet6` and `sys` missing aswell?)

Anyway currently I get:
```
generating bootimage and codeimage binaries from build/ios-arm-bootimage-openjdk-src/classpath using build/ios-arm-bootimage-openjdk-src/bootimage-generator
make: build/ios-arm-bootimage-openjdk-src/bootimage-generator: No such file or directory
```
as a result to the following call:
```
build/ios-arm-bootimage-openjdk-src/bootimage-generator -cp build/ios-arm-bootimage-openjdk-src/classpath -bootimage build/ios-arm-bootimage-openjdk-src/bootimage-bin.o -codeimage build/ios-arm-bootimage-openjdk-src/codeimage-bin.o -bootimage-symbols _binary_bootimage_bin_start:_binary_bootimage_bin_end -codeimage-symbols _binary_codeimage_bin_start:_binary_codeimage_bin_end -hostvm build/macosx-x86_64-interpret/libjvm.dylib
```
hostvm still doesn't exist, though it says `building build/ios-arm-bootimage-openjdk-src/bootimage-generator arch=x86_64 platform=macosx`.
On the other hand not even that file (ios-arm-b.../bootimage-generator) doesn't exist

**Comment by dicej** [comment]
On Fri, Aug 12, 2016 at 3:45 PM, MeFisto94 <notifications@github.com> wrote:

> I'll try to put up a test case, yes, though it's strange that you don't
> experience the same issues and that I get inconsistent results (the error
> where I'm stuck, I couldn't pass it before due to the ld: "cannot insert
> branch island", whereas arm64 yielded the error in bootimage-generation,
> now back to arm it does try to generate the bootimage)
>

"cannot insert branch island" probably means the boot image is too large,
meaning the linker can't generate an instruction to jump around it.  Make
sure you're using ProGuard to minimize the size of the code.  If the boot
image is still too large even when using ProGuard, we'll have to get
creative and figure out a workaround.


> First of all: Adding sim=true to the make command led to errors when
> trying to compile arm-asm.o on arch=asm.
> JAVA_HOME=[...]/openjdk-1.7.0-u80-unofficial-macosx-x86_64-image/ make
> arch=arm platform=ios bootimage=true process=compile
> openjdk=[...]/openjdk-1.7.0-u80-unofficial-macosx-x86_64-image/
> openjdk-src=[...]/jdk7u/jdk/src/ sim=true
>

Combining `arch=arm` and `sim=true` won't work; the simulator only
understands x86 or x86_64 code.

> I'll try to setup a test case tomorrow, I don't know how to do it
> reproducable, though. Travis would be nice, or some docker container? But
> that's a bit of an overhead.
>
A docker container would be great, but a zip file with a list of
instructions for installing the prerequisites and building it would be fine
too.  Once I have something running on my own machine, I'll be able to
debug it properly.


> Essentially the build.xml simply calls make and packages a few output
> files to a zip for each platform (arm, arm64 and i386).
>
>  <exec executable="make" dir="${avian.src.path}" failonerror="true">
>     <env key="JAVA_HOME" value="${avian.jdk.path}"/>
>     <arg line="arch=arm"/>
>     <arg line="platform=ios"/>
>     <arg line="process=compile"/>
>     <arg line="openjdk='${avian.jdk.path}'"/>
>     <arg line="openjdk-src='${avian.jdk.src.path}'"/>
>     <arg line="lzma='${avian.lzma.sdk.path}'"/>
>     <arg line="bootimage=true"/>
> </exec>
> <delete file="src/com/jme3/gde/ios/avian-openjdk-mac.zip" failonerror="false"/>
> <zip destfile="src/com/jme3/gde/ios/avian-openjdk-mac.zip">
>      <zipfileset file="${avian.src.path}/build/ios-arm-bootimage-openjdk-src/bootimage-generator" filemode="755" fullpath="avian-arm/bootimage-generator"/>
>      <zipfileset file="${avian.src.path}/build/ios-arm-bootimage-openjdk-src/binaryToObject/binaryToObject" filemode="755" fullpath="avian-arm/binaryToObject"/>
>      <zipfileset file="${avian.src.path}/build/ios-arm-bootimage-openjdk-src/libavian.a" fullpath="avian-arm/libavian.a"/>
> </zip>
>
> So it actually fails on the make part. I could work around many issues
> though (Missing GetClassLoader in jvm.h, so I took it from hotspot, missing
> netinet headers (btw, does ReadyTalk/avian#494
> <https://github.com/ReadyTalk/avian/pull/494> also take into account that
> there were some files from netinet6 and sys missing aswell?)
>

Can you send me the exact `make` command that exhibits the problem?  You
shouldn't need to copy any headers around or supplement jvm.h.  Let me know
where you got the JDK and the JDK source from.

**Comment by MeFisto94** [comment]
For the jvm.h it's the same that you mentioned on the Mailing List:
> NB: Here I ran into an issue where JVM_GetClassLoader was not
> defined in the jdk repo's jvm.h.  However, it was defined in the
> hotspot repo's jvm.h, so I just copied the declaration from there and
> re-ran make
This only happens with arm64 as target though, not on arm64.

When I execute this Script (It's slightly derived from your code) then I get:
```
wget https://bitbucket.org/alexkasko/openjdk-unofficial-builds/downloads/openjdk-1.7.0-u80-unofficial-macosx-x86_64-image.zip
unzip openjdk-1.7.0-u80-unofficial-macosx-x86_64-image.zip
hg clone http://hg.openjdk.java.net/jdk7u/jdk7u/
cd jdk7u
bash get_source.sh
cd jdk
hg checkout jdk7u80-b32
cd ../../
git clone https://github.com/readytalk/avian.git
git clone https://github.com/readytalk/hello-ios.git
curl -Of http://readytalk.github.io/avian-web/lzma920.tar.bz2
(mkdir -p lzma-920 && cd lzma-920 && tar xjf ../lzma920.tar.bz2)
curl -Of http://readytalk.github.io/avian-web/proguard4.11.tar.gz
tar xzf proguard4.11.tar.gz
cd avian
JAVA_HOME=$(pwd)/../openjdk-1.7.0-u80-unofficial-macosx-x86_64-image make arch=arm platform=ios bootimage=true process=compile openjdk=$(pwd)/../openjdk-1.7.0-u80-unofficial-macosx-x86_64-image/ openjdk-src=$(pwd)/../jdk7u/jdk/src/
```

```
Note: Recompile with -Xlint:unchecked for details.
generating bootimage and codeimage binaries from build/ios-arm-bootimage-openjdk-src/classpath using build/ios-arm-bootimage-openjdk-src/bootimage-generator
unable to open build/macosx-x86_64-interpret/libjvm.dylib
make: *** [build/ios-arm-bootimage-openjdk-src/bootimage-bin.o] Error 255
```

When I re-issue make I get (obviously):
```
linking build/ios-arm-bootimage-openjdk-src/avian
ld: file too small (length=0) file 'build/ios-arm-bootimage-openjdk-src/bootimage-bin.o' for architecture armv7
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

I hope you can reproduce that issue and it's not build system related...

**Comment by dicej** [comment]
I wasn't able to reproduce what you describe exactly, but I did run into some issues.  The problem is that running `make` with no targets specified (i.e. using the default target `build`, which tries to build everything) doesn't make sense when cross-compiling for iOS.  We can't really build a functional `avian` executable or shared library for iOS, so the makefile doesn't know how to even try.  That's why I didn't try building avian directly in the instructions I gave; instead, I built hello-ios, which in turn builds avian but only specifies the targets it needs (see the `vm-targets` variable in hello-ios/makefile).  Those include everything you need to build an iOS app (e.g. the bootimage generator, libavian.a, etc.) and nothing you don't need (e.g. libjvm.dylib, etc.).

In short, you should only build the targets listed in `vm-targets` in hello-ios/makefile when building avian for iOS; nothing else is likely to work.  Which means the makefile should probably just error out if you do try anything else, or at least do the right thing by default if you don't specify a target.

**Comment by MeFisto94** [comment]
That explains all the trouble I had from day one :smile:
Stupid me removed those build/.../libavian.a from the Make CommandLine since I thought they are outdated (I got the "no target found" error, possibly due to wrong $(options)).

Anyhow now I have troubles building for arch=i386:
```
     [exec] compiling build/ios-i386-bootimage-openjdk-src/codegen/target/x86/detect.o
     [exec] In file included from src/codegen/target/x86/detect.cpp:21:
     [exec] /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/clang/6.0/include/cpuid.h:25:2: error: this header is for x86 only
     [exec] #error this header is for x86 only
     [exec]  ^
     [exec] /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/clang/6.0/include/cpuid.h:125:5: error: invalid output constraint '=a' in asm
     [exec]     __cpuid(__level, *__eax, *__ebx, *__ecx, *__edx);
     [exec]     ^
     [exec] /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/clang/6.0/include/cpuid.h:114:21: note: expanded from macro '__cpuid'
     [exec]     __asm("cpuid" : "=a"(__eax), "=b" (__ebx), "=c"(__ecx), "=d"(__edx) \
     [exec]                     ^
     [exec] /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/clang/6.0/include/cpuid.h:153:5: error: invalid output constraint '=a' in asm
     [exec]     __cpuid(__level, __eax, __ebx, __ecx, __edx);
     [exec]     ^
     [exec] /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/clang/6.0/include/cpuid.h:114:21: note: expanded from macro '__cpuid'
     [exec]     __asm("cpuid" : "=a"(__eax), "=b" (__ebx), "=c"(__ecx), "=d"(__edx) \
     [exec]                     ^
     [exec] 3 errors generated.
     [exec] make: *** [build/ios-i386-bootimage-openjdk-src/codegen/target/x86/detect.o] Error 1
```
Seems Apple changed something? And should I rather open another issue for that, this time in the correct repo?

**Comment by dicej** [comment]
Can you show me the complete `make` command?  If you're building for x86(_64), make sure you also specify `sim=true`.

**Comment by MeFisto94** [comment]
Well it's the same as above but without sim=true. That's because that wasn't in the original Ant Task either.
Just to understand that: Are there some x86 iOS Devices out there or is that soley for the Simulator?
Anyway I was able to build x86_64 with the attached patch (you might want to include it in the repo, since I was avian is already patching the jdk sometimes).
[GetClassLoader.patch.txt](https://github.com/ReadyTalk/hello-ios/files/428787/GetClassLoader.patch.txt)

i386 doesn't work because of that: (My Host is x86_64 and for some reason it doesn't enable the 32bit mode)
```
compiling build/ios-i386-bootimage-openjdk-src/host/codegen/target/x86/detect.o
In file included from src/codegen/target/x86/detect.cpp:22:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/clang/6.0/include/cpuid.h:125:5: error: instruction requires: Not 64-bit mode
    __cpuid(__level, *__eax, *__ebx, *__ecx, *__edx);
    ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/clang/6.0/include/cpuid.h:98:11: note: expanded from macro '__cpuid'
    __asm("  pushl  %%ebx\n" \
          ^
<inline asm>:1:4: note: instantiated into assembly here
          pushl  %ebx
          ^
In file included from src/codegen/target/x86/detect.cpp:22:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/clang/6.0/include/cpuid.h:125:5: error: instruction requires: Not 64-bit mode
    __cpuid(__level, *__eax, *__ebx, *__ecx, *__edx);
    ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/clang/6.0/include/cpuid.h:101:12: note: expanded from macro '__cpuid'
          "  popl   %%ebx" \
           ^
<inline asm>:4:3: note: instantiated into assembly here
  popl   %ebx
  ^
2 errors generated.
```
With `JAVA_HOME=$(pwd)/../openjdk-1.7.0-u80-unofficial-macosx-x86_64-image make arch=i386 platform=ios bootimage=true process=compile openjdk=$(pwd)/../openjdk-1.7.0-u80-unofficial-macosx-x86_64-image/ openjdk-src=$(pwd)/../jdk7u/jdk/src/ sim=true build/ios-i386-bootimage-openjdk-src/bootimage-generator build/ios-i386-bootimage-openjdk-src/binaryToObject/binaryToObject build/ios-i386-bootimage-openjdk-src/classpath.jar build/ios-i386-bootimage-openjdk-src/libavian.a`

**Comment by dicej** [comment]
As far as I know, there are no x86(_64) iOS devices besides the simulator.  The `make` command you gave works for me, so I guess there's something different about your Xcode install compared to mine.

**Comment by MeFisto94** [comment]
Mh, searching for `instruction requires: Not 64-bit mode` shows multiple XCode 6 related issues. Unforunately I can't upgrade to XCode 7. 

https://sourceforge.net/p/jamvm/mailman/message/32910758/ recommend adding a -m32 switch, which didn't help me either (I added it to arch-flags)

I simply dropped i386 support now since everyone has an x86_64 machine to test the simulator on. Now XCode is being naughty by trying to compile an i386 version even if I didn't select it in the archs.
I might really need to update my OSX.

Anyway, the games are now running again thanks to your awesome support about me being to stupid to issue make correctly. Thank you guys!
